### PR TITLE
ENH: Re-enable analysis level selection

### LINKED ISSUE #102: Question: What to do with deprecated "analysis_level" as required input
Ever since I started to use fitlins I've been confused why this required [`analysis_level`](https://github.com/poldracklab/fitlins/blob/c7cc798330a8fe8cea8ee545c64c7fb08d1a562b/fitlins/cli/run.py#L70) argument is (still) in existence: https://github.com/poldracklab/fitlins/blob/c7cc798330a8fe8cea8ee545c64c7fb08d1a562b/fitlins/cli/run.py#L70-L71

I believe it has been deprecated half a year ago, and I can't find it being used anywhere, but running fitlins without it leads to

```shell
usage: fitlins [-h] [-v]
               [--participant-label PARTICIPANT_LABEL [PARTICIPANT_LABEL ...]]
               [-m MODEL] [-d DERIVATIVES [DERIVATIVES ...]]
               [--derivative-label DERIVATIVE_LABEL]
               [--space {MNI152NLin2009cAsym,,MNI152NLin6Sym}]
               [--include INCLUDE] [--exclude EXCLUDE] [--desc DESC]
               [--n-cpus N_CPUS] [--debug] [-w WORK_DIR]
               bids_dir output_dir {run,session,participant,dataset}
fitlins: error: the following arguments are required: analysis_level
```

I could try to contribute and make analysis_level a 'stopping criterion' to halt the computations once the specified analysis level is reached, if that is intended/useful functionality. Otherwise I could just remove it from argparse.

### PR DESCRIPTION
**Pull request by effigies**
cc @mgxd

Closes #102.

### CODE REVIEW

File: fitlins/cli/run.py [hunk 1]
```diff
@@ -170,9 +170,9 @@ def run_fitlins(argv=None):
     work_dir = mkdtemp() if opts.work_dir is None else opts.work_dir
 
     fitlins_wf = init_fitlins_wf(
-        opts.bids_dir, derivatives, deriv_dir, opts.space,
-        desc=opts.desc_label,
-        model=model,
+        opts.bids_dir, derivatives, deriv_dir,
+        analysis_level, model=model,
```
Comment (mgxd) [inline fitlins/cli/run.py]:
```suggestion
        analysis_level=opts.analysis_level, model=model,
```

File: fitlins/interfaces/bids.py [hunk 2]
```diff
@@ -182,10 +184,20 @@ def _run_interface(self, runtime):
 
         selectors = self.inputs.selectors
 
+        if analysis_level == 'participant':
```
Comment (mgxd) [inline fitlins/interfaces/bids.py]:
```suggestion
        if self.inputs.analysis_level == 'participant':
```

File: fitlins/interfaces/bids.py [hunk 3]
```diff
@@ -182,10 +184,20 @@ def _run_interface(self, runtime):
...
+            analysis_level = 'subject'
```
Comment (mgxd) [inline fitlins/interfaces/bids.py]:
```suggestion
            self.inputs.analysis_level = 'subject'
```

File: fitlins/interfaces/bids.py [hunk 4]
```diff
@@ -182,10 +184,20 @@ def _run_interface(self, runtime):
...
+
+        steps = []
+        for i, step in enumerate(model['Steps']):
```
Comment (mgxd) [inline fitlins/interfaces/bids.py]:
```suggestion
        for i, step in enumerate(self.inputs.model['Steps']):
```

File: fitlins/interfaces/bids.py [hunk 5]
```diff
@@ -182,10 +184,20 @@ def _run_interface(self, runtime):
...
+            if step['Level'] == analysis_level:
+                break
+
         analysis = Analysis(model=self.inputs.model, layout=layout)
-        analysis.setup(drop_na=False, desc='preproc', **selectors)
+        analysis.setup(steps=steps, drop_na=False, desc='preproc', **selectors)
         self._load_level1(runtime, analysis)
-        self._load_higher_level(runtime, analysis)
+        if len(steps) > 1:
+            self._load_higher_level(runtime, analysis, analysis_level)
```
Comment (mgxd) [inline fitlins/interfaces/bids.py]:
```suggestion
            self._load_higher_level(runtime, analysis, self.inputs.analysis_level)
```

**Comment by mgxd** [comment]
I'm still running into errors - will update shortly.

BTW - can I suggest setting the workflow config to output crashfiles as `.txt` files? It makes debugging much easier, especially across containers.

**Comment by mgxd** [comment]
<details>
<summary>Crash 1</summary>
Node: fitlins_wf.select_l4_contrasts
Working directory: /working/fitlins_wf/select_l4_contrasts

Node inputs:

index = [3]
inlist = [[[{'entities': {'run': 1, 'session': '1', 'subject': 'voice969', 'task': 'emosent'}, 'name': 'speech', 'type': 't', 'weights': [{'trial_type.happy': 0.33, 'trial_type.neutral': 0.33, 'trial_type.sad': 0.33}]}], [{'entities': {'run': 2, 'session': '1', 'subject': 'voice969', 'task': 'emosent'}, 'name': 'speech', 'type': 't', 'weights': [{'trial_type.happy': 0.33, 'trial_type.neutral': 0.33, 'trial_type.sad': 0.33}]}]], [[{'entities': {'session': '1', 'subject': 'voice969', 'task': 'emosent'}, 'name': 'speech', 'type': 't', 'weights': [{'speech': 1}]}]], [[{'entities': {'session': '1', 'subject': 'voice969', 'task': 'emosent'}, 'name': 'speech', 'type': 't', 'weights': [{'speech': 1}]}]]]

Traceback (most recent call last):
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 316, in _send_procs_to_workers
    self.procs[jobid].run(updatehash=updatehash)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 473, in run
    result = self._run_interface(execute=True)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 557, in _run_interface
    return self._run_command(execute)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 637, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 377, in run
    outputs = self.aggregate_outputs(runtime)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 455, in aggregate_outputs
    predicted_outputs = self._list_outputs()
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/interfaces/utility/base.py", line 404, in _list_outputs
    self.inputs.index)].tolist()
IndexError: index 3 is out of bounds for axis 1 with size 3
</details>
<details>
<summary>Crash 2</summary>
Node: fitlins_wf.l3_model
Working directory: /working/fitlins_wf/l3_model

Node inputs:

contrast_info = [[{'entities': {'session': '1', 'subject': 'voice969', 'task': 'emosent'}, 'name': 'speech', 'type': 't', 'weights': [{'speech': 1}]}]]
stat_files = <undefined>
stat_metadata = <undefined>

Traceback (most recent call last):
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 69, in run_node
    result['result'] = node.run(updatehash=updatehash)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 473, in run
    result = self._run_interface(execute=True)
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 1253, in _run_interface
    self.config['execution']['stop_on_first_crash'])))
  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 1175, in _collate_results
    (self.name, '\n'.join(msg)))
Exception: Subnodes of node: l3_model failed:
Subnode 0 failed
Error: Traceback (most recent call last):

  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 99, in nodelist_runner
    result = node.run(updatehash=updatehash)

  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 473, in run
    result = self._run_interface(execute=True)

  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 557, in _run_interface
    return self._run_command(execute)

  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 637, in _run_command
    result = self._interface.run(cwd=outdir)

  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 375, in run
    runtime = self._run_interface(runtime)

  File "/src/fitlins/fitlins/interfaces/nistats.py", line 178, in _run_interface
    model.fit(input, design_matrix=design_matrix)

  File "/opt/miniconda-latest/envs/neuro/lib/python3.6/site-packages/nistats/second_level_model.py", line 170, in fit
    raise ValueError('A second level model requires a list with at'

ValueError: A second level model requires a list with atleast two first level models or niimgs]
</details>

<details>
<summary>Command</summary>
fitlins --participant-label voice969 -m /data/models/model-sparse_desc-emosent_smdl.json -w /working /data /data/derivatives/fitlins participant
</details>

**Comment by effigies** [comment]
Thanks for those. Sorry about the total failure to run a style checker that would have caught all (or at least most) of them. Ran it now.

**Comment by effigies** [comment]
Ah right. Pushed a fix. I think I may want to consider reverting all of the changes in the BIDS interface and simply restrict the workflow build-out. But let's make sure this works before we go forward.

**Comment by mgxd** [comment]
Ok - things have run through successfully! Now I just need to understand why the images look so bad :disappointed:

**Comment by effigies** [comment]
Okay. Can you try again with the latest to make sure that I'm not breaking anything, then?

**Comment by mgxd** [comment]
No difference, so it's probably something with my design - I think you can push back the changes.

**Pull request merged**